### ISSUE TITLE
[ENH] generate_report method should also work after fit_transform was called

### ISSUE THREAD

**Issue by JohannesWiesner** [nilearn/nilearn#3709]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe your proposed enhancement in detail.

Currently, it seems to be the case that the `.generate_report` method of masker objects can only be called with a previous call of `.fit()` but not when `.fit_transform()` was called (which is contra-intuitive to me because `.fit_transform()` is just `.fit()` + `.transform()`). Most of the time, users will probably use `.fit_transform()` to get their numpy arrays. If they then want to check if the masking worked they currently have to call `.fit()` one more time to be able to call  `.generate_report`.

### Benefits to the change

Avoids calling `masker.fit()` one more time if a report should be generated. 

### Pseudocode for the new behavior, if applicable

MRE:

```python
from nilearn import datasets
from nilearn.maskers import NiftiLabelsMasker
import numpy as np
from nilearn.image import load_img

# download a single statistical image
motor_images = datasets.fetch_neurovault_motor_task()
stat_img = motor_images.images[0]

# get an atlas
atlas_data = datasets.fetch_atlas_schaefer_2018()

# get the atlas image. This has 401 values, with the 0 included
atlas_img = load_img(atlas_data['maps'])
print(len(np.unique(atlas_img.get_fdata())))

# we have to insert a 0 here so that img and labels have the same number of value
# This is already reported, see #3708
atlas_labels = atlas_data.labels.tolist()
atlas_labels.insert(0,'Background') 

# set a masker
masker = NiftiLabelsMasker(labels_img=atlas_img,labels=atlas_labels)

```
**Would be nice if this would work!**

```python

data = masker.fit_transform(stat_img)
report = masker.generate_report()
report.open_in_browser()
```

**But we have to call .fit() one more time in order to be able to generate a report**

```
masker.fit(stat_img)
report = masker.generate_report()
report.open_in_browser()
```

**Comment by ymzayek** [comment]
Hi @JohannesWiesner, just to clarify the issue...

You are saying that the behavior of the report generator should be the same whether you:

1. First call `fit(stat_img)`, then `transform(stat_img)`, and lastly `generate_report()`

or if you: 

2. Call `fit_transform(stat_img)` and then `generate_report()`

However, it is not the same because the results from 1 are

![image](https://user-images.githubusercontent.com/63292494/234289077-1efa0eab-8785-4c61-92f6-d12a2b24c375.png)

And from 2:

![image](https://user-images.githubusercontent.com/63292494/234289213-aba289e4-b8d2-4abe-9518-a027cabe39c2.png)

Is that accurate?

This is because  `fit_transform` first calls `fit()` without passing the img data and then calls `transform`. So it's the same as 1. if you didn't pass stat_img to `fit`.

I'm not sure this is a bug or should be changed because `fit` is where certain parameters are set that are passed to the reporter so it's expected behavior but I understand it could be confusing without looking at the code. I'd be interested in hearing the opinion of others on this.

**Comment by bthirion** [comment]
I think that the first option with a better colormap would be the expected the right result. and I agree that calling `fit_transform()` should match the call of `fit()` and `transform()` with the same arguments.

**Comment by ymzayek** [comment]
Closing this in favor of issue #3933 since the other things in this issue have been addressed by #3897
